src/rgw/driver/rados/rgw_rados.cc

```diff
@@ -10063,6 +10063,10 @@ int RGWRados::check_bucket_shards(const RGWBucketInfo& bucket_info,
       return 0;
   }
 
+  if (! is_layout_reshardable(bucket_info.layout)) {
+    return 0;
+  }
+
```

Comment:
i'm not sure 'is reshardable' is a useful abstraction here:
* it doesn't save us from using the asserting version of `rgw::current_num_shards()` below
* different sharding schemes will have different constraints about sizing, so we're unlikely to share the math based on `rgw_max_objs_per_shard`

i think it's fine for this function to be specific to `BucketIndexType::Normal`. we just don't want it to crash on other types

Reply:
The code is now written with multiple layouts in mind, and it sounds like there is a 3rd layout in the works. It seems a high-level test of reshard-ability is helpful and makes the code more readable to those who may not know the details of each layout type.

Reply:
i like `is_layout_reshardable()` better now that it's used in places other than just `RGWRados::check_bucket_shards()`. but the indexless crash came from the asserting version of `current_num_shards(BucketLayout)` which this still doesn't replace

the only way to avoid the asserting overload is to use a specific layout type like `num_shards(bucket_index_normal_layout)`

i'm leaving my approval because this does fix the indexless crash. we won't hit this assertion unless `is_layout_reshardable()` and `current_num_shards()` diverge in the future